File: src/types/coaching.ts
export type TipPriority = 'info' | 'warning' | 'danger';

export type TipCategory =
  | 'farming'
  | 'trading'
  | 'vision'
  | 'objective'
  | 'positioning'
  | 'recall'
  | 'matchup'
  | 'jungle'
  | 'teamfight'
  | 'macro'
  | 'general';

export interface CoachingTip {
  id: string;
  message: string;
  priority: TipPriority;
  category: TipCategory;
  timestamp: number;
  dismissed: boolean;
}

export interface MatchupInfo {
  yourChampion: string;
  enemyChampion: string;
  difficulty: 1 | 2 | 3 | 4 | 5;
  tips: string[];
  powerSpikes: string[];
  laneSummary: string;
}

export type ThreatLevel = 'low' | 'medium' | 'high';

export type ThreatType =
  | 'burst'
  | 'sustained'
  | 'cc'
  | 'poke'
  | 'engage'
  | 'splitpush'
  | 'assassin'
  | 'dive';

export interface ThreatAssessment {
  championName: string;
  threatLevel: ThreatLevel;
  threatTypes: ThreatType[];
  notes: string;
}

export type JungleSide = 'top' | 'mid' | 'bot';

export interface JunglePrediction {
  predictedSide: JungleSide;
  confidence: number;
  lastSeen: number;
  gankRisk: number;
}

export interface FarmingStats {
  currentCS: number;
  csPerMin: number;
  targetCS: number;
  missedCS: number;
  efficiency: number;
}

export type ObjectiveType = 'dragon' | 'baron' | 'herald';
export type ObjectiveStatus = 'alive' | 'dead' | 'spawning';

export interface ObjectiveInfo {
  type: ObjectiveType;
  status: ObjectiveStatus;
  timer: number;
  dragonType?: string;
}

export type PerformanceGrade = 'S+' | 'S' | 'A' | 'B' | 'C' | 'D' | 'F';

export interface PostGameGrades {
  laning: PerformanceGrade;
  farming: PerformanceGrade;
  vision: PerformanceGrade;
  teamfighting: PerformanceGrade;
  objectives: PerformanceGrade;
  overall: PerformanceGrade;
}

export interface KeyMistake {
  timestamp: number;
  description: string;
  category: TipCategory;
  severity: TipPriority;
}

export interface PostGameStats {
  kills: number;
  deaths: number;
  assists: number;
  cs: number;
  csPerMin: number;
  visionScore: number;
  goldEarned: number;
  damageDealt: number;
  killParticipation: number;
}

export interface PostGameAnalysis {
  grades: PostGameGrades;
  keyMistakes: KeyMistake[];
  stats: PostGameStats;
  improvementTips: string[];
  duration: number;
}

export type AIStatusType = 'idle' | 'thinking' | 'analyzing' | 'coaching';

export interface AIState {
  status: AIStatusType;
  currentThought: string;
  reasoningChain: string[];
}

export type GamePhase =
  | 'WAITING'
  | 'LOADING'
  | 'EARLY_GAME'
  | 'MID_GAME'
  | 'LATE_GAME'
  | 'POST_GAME';

File: backend/models/coaching.py
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class Priority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class CoachingTip(BaseModel):
    message: str
    priority: Priority = Priority.MEDIUM
    category: str = "general"


class CoachingRequest(BaseModel):
    game_state: dict[str, Any]
    player_name: str
    game_time: float


class CoachingResponse(BaseModel):
    tips: list[CoachingTip] = Field(default_factory=list)
    reasoning: str = ""
    ai_state: str = "ready"


class MatchupRequest(BaseModel):
    your_champion: str
    enemy_champion: str
    lane: str
    game_time: float = 0.0


class MatchupResponse(BaseModel):
    difficulty: int = 3
    tips: list[str] = Field(default_factory=list)
    power_spikes: dict[str, list[str]] = Field(default_factory=dict)
    lane_summary: str = ""


class JunglePredictionRequest(BaseModel):
    game_state: dict[str, Any]
    events: list[dict[str, Any]] = Field(default_factory=list)


class JunglePredictionResponse(BaseModel):
    predicted_side: str = "unknown"
    confidence: float = 0.0
    gank_risk: dict[str, float] = Field(default_factory=dict)
    reasoning: str = ""


class GradeBreakdown(BaseModel):
    laning: str = "B"
    farming: str = "B"
    fighting: str = "B"
    vision: str = "B"
    objectives: str = "B"
    overall: str = "B"


class PostGameRequest(BaseModel):
    game_data: dict[str, Any]
    events_history: list[dict[str, Any]] = Field(default_factory=list)


class PostGameResponse(BaseModel):
    grades: GradeBreakdown = Field(default_factory=GradeBreakdown)
    key_mistakes: list[str] = Field(default_factory=list)
    improvement_tips: list[str] = Field(default_factory=list)
    stats: dict[str, Any] = Field(default_factory=dict)
    summary: str = ""
